backend/core/modules/auth/service.go

```diff
@@ -0,0 +1,132 @@
+package auth
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"Tella-Desktop/backend/utils/authutils"
+	"Tella-Desktop/backend/utils/constants"
+
+	"github.com/wailsapp/wails/v2/pkg/runtime"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+type service struct {
+	ctx          context.Context
+	tvaultPath   string
+	databasePath string
+	databaseKey  []byte
+	isUnlocked   bool
+}
+
+func NewService(ctx context.Context) Service {
+	return &service{
+		ctx:          ctx,
+		tvaultPath:   authutils.GetTVaultPath(),
+		databasePath: authutils.GetDatabasePath(),
+		isUnlocked:   false,
+	}
+}
+
+func (s *service) Initialize(ctx context.Context) error {
+	s.ctx = ctx
+
+	// create directory if they don't exists
+	vaultDir := filepath.Dir(s.tvaultPath)
+	if err := os.MkdirAll(vaultDir, 0755); err != nil {
+		return fmt.Errorf("failed to create vault directory: %w", err)
+	}
+
+	// create tmp directory for decrypted files
+	tempDir := authutils.GetTempDir()
+	if err := os.MkdirAll(tempDir, 0755); err != nil {
+		return fmt.Errorf("failed to create temp directory: %w", err)
+	}
+
+	runtime.LogInfo(ctx, "Auth service initialized")
+	return nil
+}
+
+func (s *service) IsFirstTimeSetup() bool {
+	// Check if the tvault file exists
+	_, err := os.Stat(s.tvaultPath)
+	return os.IsNotExist(err)
+}
+
+func (s *service) CreatePassword(password string) error {
+	if len(password) < 6 {
+		return constants.ErrPasswordTooShort
+	}
+
+	//generate random database key | TODO: move this outside of this function
+	dbKey := make([]byte, constants.KeyLength)
+	if _, err := rand.Read(dbKey); err != nil {
+		return fmt.Errorf("failed to generate database key: %w", err)
+	}
+
+	//generate random salt
+	salt := make([]byte, constants.SaltLength)
+	if _, err := rand.Read(salt); err != nil {
+		return fmt.Errorf("failed to generate salt: %w", err)
+	}
+
+	//derive key from password and salt
+	derivedKey := pbkdf2.Key([]byte(password), salt, constants.Iterations, constants.KeyLength, sha256.New)
+
+	//encrypt database key using derived key
+	encryptedDBKey, err := authutils.EncryptData(dbKey, derivedKey)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt database key: %w", err)
+	}
+
+	//create and write tvault header
+	if err := authutils.WriteTVaultHeader(salt, encryptedDBKey); err != nil {
+		return fmt.Errorf("failed to write tvault header: %w", err)
+	}
+
+	// Store database key in memory
+	s.databaseKey = dbKey
+	s.isUnlocked = true
+
+	runtime.LogInfo(s.ctx, "Password created successfully")
+	return nil
+}
+
+func (s *service) VerifyPassword(password string) ([]byte, error) {
```

Comment:
suggestion: refactor/rename: `VerifyPassword(password string) ([]byte, error)`. for example a more instructive name could be `DecryptDatabaseKey(password string) ([]byte, error)`.

also a question: why return the decrypted key if it's also accessible on `(s Service)` via `s.databaseKey`?

Reply:
Yeah, this makes total sense! 

For the naming part, I changed the service method to `DecryptDatabaseKey(password string) error` but I kept the method in the app.go that the frontend calls as`VerifyPassword`. Conceptually, the user is verifying their password to authenticate, even though the function is actually decrypting the db key. I thought it was more intuitive to keep it that way but let me know what do you think!

And yes, it is redundant to return the db key so I removed that return, thanks for catching it!

Reply:
> For the naming part, I changed the service method to DecryptDatabaseKey(password string) error 

this is a more descriptive name, i like it! :>